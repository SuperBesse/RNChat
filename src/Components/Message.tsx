import React from 'react';
import { Text, View, StyleSheet, TouchableOpacity } from 'react-native';
import type { Message as IMEssage } from '@/Types/Message';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { faCircleExclamation } from '@fortawesome/free-solid-svg-icons/faCircleExclamation';
import { useDispatch } from 'react-redux';
import { resendMessage } from '@/Redux/Actions/MessageActions';

type Props = {
  message: IMEssage;
};

const Message = ({ message }: Props) => {
  const formattedDate = (timestamp: number): string => {
    const timestampToDate = new Date(timestamp);
    return timestampToDate.toLocaleTimeString([], {
      hour: '2-digit',
      minute: '2-digit',
    });
  };

  const dispatch = useDispatch();

  const reSendMessage = () => {
    dispatch(resendMessage(message));
  };

  const { notSent } = message;
  return (
    <TouchableOpacity
      testID={`testid-${message.date}`}
      style={styles.container}
      disabled={!notSent}
      onPress={reSendMessage}
    >
      <View style={styles.dateContainer}>
        <Text style={styles.date}>{formattedDate(message.date)}</Text>
        {notSent && (
          <FontAwesomeIcon
            style={styles.errorIcon}
            icon={faCircleExclamation}
          />
        )}
      </View>
      <View style={styles.messageContainer}>
        <Text style={styles.content}>{message.content}</Text>
      </View>
    </TouchableOpacity>
  );
};

export default Message;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'flex-end',
    paddingHorizontal: 20,
    marginVertical: 5,
  },
  dateContainer: {
    marginRight: 6,
  },
  errorIcon: {
    color: '#FF0000EE',
    alignSelf: 'flex-end',
  },
  messageContainer: {
    borderTopLeftRadius: 12,
    borderBottomLeftRadius: 12,
    borderBottomRightRadius: 12,
    paddingHorizontal: 10,
    paddingVertical: 12,
    maxWidth: '60%',
    backgroundColor: '#4b9cba',
  },
  content: {
    fontSize: 14,
  },
  date: {
    marginVertical: 4,
    fontSize: 10,
  },
});
