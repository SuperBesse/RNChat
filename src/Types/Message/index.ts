export type Message = {
  content: string;
  date: number;
  notSent: boolean;
};
