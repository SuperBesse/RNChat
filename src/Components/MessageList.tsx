import React from 'react';
import { FlatList, ViewStyle, ListRenderItem, StyleSheet } from 'react-native';
import Message from './Message';
import { Message as IMessage } from '@/Types/Message';
import { useSelector } from 'react-redux';
import { AppState } from '@/Redux/Reducers';

type Props = {
  style?: ViewStyle;
};

const MessageList = ({ style }: Props) => {
  const renderItem: ListRenderItem<IMessage> = ({ item }) => (
    <Message message={item} />
  );

  const messages = useSelector(
    (state: AppState) => state.messagesState.messages,
  );

  return (
    <FlatList
      style={[style, styles.list]}
      contentContainerStyle={style}
      data={messages}
      renderItem={renderItem}
      keyExtractor={(item: IMessage) => item.date.toString()}
      persistentScrollbar
      inverted
    />
  );
};

export default MessageList;

const styles = StyleSheet.create({
  list: {
    flex: 1,
  },
});
MessageList.defaultProps = {
  style: undefined,
};
