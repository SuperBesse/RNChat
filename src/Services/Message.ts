import { Message } from '@/Types/Message';

const randomTime = () => Math.floor(Math.random() * 2000);

const delay = () => new Promise(resolve => setTimeout(resolve, randomTime()));

const randomError = (): Boolean => Math.floor(Math.random() * 100) < 30;

const sendMessage = async (message: Message): Promise<boolean> => {
  console.log('Send Message API: ', message);
  const fakeError: Boolean = randomError();

  if (fakeError) {
    throw new Error('fake error');
  }
  await delay();
  return fakeError;
};

export default sendMessage;
