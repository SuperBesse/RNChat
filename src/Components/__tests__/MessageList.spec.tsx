import React from 'react';
import { render, screen } from '@testing-library/react-native';
import MessageList from '../MessageList';
import { FlatList } from 'react-native';
import Message from '../Message';
import type { Message as IMEssage } from '@/Types/Message';
import * as Redux from 'react-redux';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

describe('MessageList', () => {
  it('should render correctly', () => {
    const tree = render(<MessageList />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(screen.UNSAFE_getAllByType(FlatList).length).toBe(1);
  });

  it('should render message when data is updated', () => {
    render(<MessageList />).toJSON();

    const message = {
      content: 'message 1',
      date: 1487076708000,
      notSent: true,
    } as IMEssage;
    jest.spyOn(Redux, 'useSelector').mockReturnValue([message]);
    screen.rerender(<MessageList />);
    expect(screen.UNSAFE_getAllByType(Message).length).toBe(1);
  });
});
