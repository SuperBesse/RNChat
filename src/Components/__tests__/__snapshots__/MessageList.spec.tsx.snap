// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList should render correctly 1`] = `
<RCTScrollView
  getItem={[Function]}
  getItemCount={[Function]}
  invertStickyHeaders={true}
  inverted={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  persistentScrollbar={true}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      Array [
        undefined,
        Object {
          "flex": 1,
        },
      ],
    ]
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View />
</RCTScrollView>
`;
